### DISCUSSION TITLE
How do json_patches work in practice?

### DISCUSSION THREAD

**Discussion by joekale** [pboettch/json-schema-validator#200] [Q&A]
I have been learning to build schemas and am trying to apply them to a current project. I have been running into issues with patching that I believe to be a misunderstanding, but want to be sure.

Is my understanding correct, that when a required property is missing, and there is a default defined for the property, I should receive back a json_patch object that can be used to bring the json under validation up to schema spec?

If I am correct, are there limitations to this? Can my schema define a default that gets provided back in the case of an empty object? Currently when I try to validate an empty object I receive an empty object for the returned patch.

**Comment by pboettch** [comment]
> I have been learning to build schemas and am trying to apply them to a current project. I have been running into issues with patching that I believe to be a misunderstanding, but want to be sure.
> 
> Is my understanding correct, that when a required property is missing, and there is a default defined for the property, I should receive back a json_patch object that can be used to bring the json under validation up to schema spec?
> 
> If I am correct, are there limitations to this? Can my schema define a default that gets provided back in the case of an empty object? Currently when I try to validate an empty object I receive an empty object for the returned patch.

Please provide an example of your schema and instance and what you expect.

**Reply by joekale** [reply]
I worked up a simple schema that I think highlights my current dilemma.
```
{
    "$schema": "http://json-schema.org/draft-07/schema#",
    "type": "object",
    "required": ["foo", "bar"],
    "properties": {
      "foo": {
       	"test": "array",
        "items": {
          "$ref": "#/definitions/foo"
        },
        "default": [
          {
            "$ref": "#/definitions/foo"
          }
         ]
      },
      "bar": {
        "type": "object",
       	"oneOf": [
          {
            "$ref": "#/definitions/bar1"
          },
          {
            "$ref": "#/definitions/bar2"
          }
        ],
        "default": {
          "$ref": "#/definitions/bar1"
        }
      }
      
    },
    "definitions": {
      "foo": {
      	"type": "object",
        "properties": {
          "foo1": {
            "type": "string",
            "default": "test"
          }
        }
      },
      "bar1": {
       	"type": "object",
        "properties": {
          "bar1": {
            "type": "string",
            "const": "bar1"
          }
        },
        "required": ["bar1"]
      },
      "bar2": {
       	"type": "object",
        "properties": {
          "bar2": {
            "type": "string",
            "const": "bar2"
          }
        },
        "required": ["bar2"]
      }
    },
    "default": {
      "foo": {
       	"foo1": "top_def" 
      },
      "bar": {
        "$ref": "#/definitions/bar2"
      }
    }
}
```
With that schema I have a couple of scenarios. For the first ignore the top level defined "default". If I am provided an empty object `{}` or an object with keys that still don't validate against the schema, I would expect that a patch would be created to add `foo` with a single foo element that is a string with value "test", and add `bar` with a `bar1` object.

For the second scenario, if I am provided an empty object and I have the top level defined "default" I would assume that I would be returned the full default object containing `foo` with a string being `top_def` and a bar with a `bar2` object. I do not know if this scenario is in spec to the intent of patching though.

**Reply by pboettch** [reply]
Just a quick reply, a default-field should be a value not a schema. As such, if a default-field is not evaluated as a schema and $ref's are not resolved.

I never worked with default-value-patches personally but IIRC the output JSON-patch will only contain formerly empty fields where the schema contains a default-value and not ones which are not validating. I'm also not sure how duplicated default-values are implemented. 

We should extend the test-cases for this functionality.

How a validator is handling default-values is not described in the standard (at least not in draft-7).

**Reply by joekale** [reply]
I believe this makes sense to me. Thank you for the response. I will try and work more with this and craft my schema in a better way to support default-value-patches.
